core/src/main/java/org/apache/spark/shuffle/sort/ShuffleInMemorySorter.java

```diff
@@ -69,7 +72,11 @@ public int numRecords() {
     return pos;
   }
 
-  public void reset() {
+  public void freeMemoryAndResetPointerArray() {
+    if (consumer != null) {
+      consumer.freeArray(array);
+      this.array = consumer.allocateArray(initialSize);
```

Comment:
There is very small chance that this could fail, should we grab a lock from TaskMemoryManager for these two?